### PR TITLE
[LoRA] Converter protocol and PEFT external format support

### PR DESCRIPTION
**Pull request by mori360**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #3189
* #3188

Add converter protocol for external format transforms (e.g. PEFT key remapping). Converters are stored on ModelSpec.converters and wired via BaseModel.set_sd_transforms(sd_adapter, converters).

ModelSpec:
- Add converters: list[Any] field. Config registry appends converter instances after convert(). Trainer iterates via duck typing.

BaseModel:
- Extend set_sd_transforms to accept converters list. Each converter may implement build_external_transforms(sd_adapter) returning {to_external, from_external} callables.
- to_external/from_external now route by mode: BASE applies HF remapping, TRAINABLE applies converter transforms, FULL applies both.

LoRAConverter:
- Add build_external_transforms instance method returning PEFT key remapping (remap_lora_keys_to_hf / remap_lora_keys_from_hf).
- Remove _is_adapter / _key_pattern ClassVars (filtering is by requires_grad, not pattern matching).

Config registry:
- llama3_debugmodel_lora: append converter to model_spec.converters.
- llama3_debugmodel_lora_checkpoint: LoRA with HF save enabled.

LoRA TP Determinism with Seed Checkpoint:
With seed checkpoint and non-chunked CrossEntropyLoss, LoRA training produces bitwise identical loss and grad_norm across TP=1/2/4, DP=2 over 20 steps:
```
  LoRA + CrossEntropyLoss + seed checkpoint: TP=1 vs TP=2 vs TP=4 

  Loss:
  Step  1: TP1=8.02020  TP2=8.02021  TP4=8.02018
  Step  5: TP1=7.97558  TP2=7.97558  TP4=7.97557
  Step 10: TP1=7.72779  TP2=7.72779  TP4=7.72779
  Step 15: TP1=7.42778  TP2=7.42787  TP4=7.42784
  Step 20: TP1=7.22069  TP2=7.22079  TP4=7.22076

  Grad norm:
  Step  1: TP1=0.1591   TP2=0.1591   TP4=0.1592
  Step  5: TP1=0.2866   TP2=0.2866   TP4=0.2866
  Step 10: TP1=0.6532   TP2=0.6533   TP4=0.6533
  Step 15: TP1=1.1474   TP2=1.1473   TP4=1.1474
  Step 20: TP1=1.4419   TP2=1.4417   TP4=1.4417
```
Loss and grad_norm match to 5 significant digits (the stdout display precision) with no drift over 20 steps. Multiple steps show exact zero diff. This confirms that LoRA's TP parallelization — adapter sharding, frozen base weights, and gradient computation — is numerically correct across TP degrees.                                                                                                               
The seed checkpoint ensures identical initial weights (base + LoRA adapters) across all TP configurations, eliminating data RNG divergence as a variable. The non-chunked CrossEntropyLoss avoids a known ChunkedCELoss cross-TP determinism issue.

### CODE REVIEW

File: torchtitan/components/lora.py [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+    from single-GPU by a constant factor (TP degree). This is a known
+    DTensor ``_NormPartial`` norm computation issue with mixed
+    Replicate/Shard placements, not a LoRA correctness bug — the loss
+    and convergence are identical.
+    """
+    replicate_weight = ShardingConfig(
+        state_shardings={"weight": dense_param_placement(tp=Replicate())},
+    )
+    if base_sharding is None:
+        return replicate_weight, replicate_weight
```
Comment (tianyu-l) [inline torchtitan/components/lora.py]: can we ban this case? When would we need LoRA sharding when base sharding is None?

File: torchtitan/components/lora.py [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+
+    base_weight_sharding = base_sharding.state_shardings.get("weight")
+    if base_weight_sharding is None:
+        return replicate_weight, replicate_weight
```
Comment (tianyu-l) [inline torchtitan/components/lora.py]: same here

File: torchtitan/components/lora.py [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+    ``base_out``'s placement (Shard(-1) for colwise, Partial for rowwise).
+
+    ``lora_a`` weight is Replicate (small rank dim, no benefit from TP
+    sharding).
+
+    Note: under TP, ``clip_grad_norm_`` reports a grad_norm that differs
+    from single-GPU by a constant factor (TP degree). This is a known
+    DTensor ``_NormPartial`` norm computation issue with mixed
+    Replicate/Shard placements, not a LoRA correctness bug — the loss
+    and convergence are identical.
```
Comment (tianyu-l) [inline torchtitan/components/lora.py]: Is there an issue tracking this?

File: torchtitan/components/lora.py [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+        return _lora_class_cache[parent_cls]
+
+    class LoRALinear(parent_cls):  # type: ignore[valid-type, misc]
+        def __init__(
+            self,
+            config: Linear.Config,
+            *,
+            _lora_rank: int,
+            _lora_alpha: float,
+            _base_sharding: ShardingConfig | None = None,
```
Comment (tianyu-l) [inline torchtitan/components/lora.py]: this should be available in config above, so don't need to pass separately

File: torchtitan/components/lora.py [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+            else:
+                wrapped = FrozenConfig(inner=cfg)
+
+            if isinstance(parent, list):
+                parent[attr] = wrapped
+            else:
+                setattr(parent, attr, wrapped)
+
+        # Freeze non-Linear leaf modules (norms, embeddings)
+        for config_cls in (RMSNorm.Config, Embedding.Config):
```
Comment (tianyu-l) [inline torchtitan/components/lora.py:266]: This doesn't sound robust. A better way is to traverse all configs in one pass and freeze anything that's not Linear.

File: torchtitan/components/lora.py [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+            config=replace(inner),
+            _lora_rank=self.rank,
+            _lora_alpha=self.alpha,
+            _base_sharding=inner.sharding_config,
+        )
+        # Apply Module.Config properties that inner.build() would set
+        if inner.param_init is not None:
+            instance._param_init = inner.param_init
+        if inner.sharding_config is not None:
+            instance._sharding_config = inner.sharding_config
```
Comment (tianyu-l) [inline torchtitan/components/lora.py]: I think the reason you are using this is because you didn't have a Config of `lora_cls` so you are not using the `build` function here https://github.com/pytorch/torchtitan/blob/main/torchtitan/protocols/module.py#L47 which does these for you.

We should have a Config inside `class LoRALinear(parent_cls)` and put rank and alpha in config.

File: torchtitan/components/lora.py [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+        if name in type(self).__slots__:
+            object.__setattr__(self, name, value)
+        else:
+            setattr(self.inner, name, value)
+
+    def build(self, **kwargs):
+        from dataclasses import replace
+
+        inner = self.inner
+        assert inner._owner is not None
```
Comment (tianyu-l) [inline torchtitan/components/lora.py:129]: I don't think it would work with today's mxfp8 linear, whose config doesn't have an `_owner` https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/quantization/mx.py#L158

I think a more consistent way is to always dynamically create a class for mxfp8, similar to the LoRALinear in this PR.
cc @acisseJZhong

File: torchtitan/models/llama3/config_registry.py [hunk 8]
```diff
@@ -104,6 +104,28 @@ def llama3_debugmodel_float8() -> Trainer.Config:
...
+def llama3_debugmodel_lora() -> Trainer.Config:
+    from torchtitan.components.lora import LoRAConverter
+
+    config = llama3_debugmodel()
+    lora_converter = LoRAConverter(
+        LoRAConverter.Config(rank=8, alpha=16.0, target_modules=["wq", "wkv", "wo"])
+    )
+    assert config.model_spec is not None
+    lora_converter.convert(config.model_spec.model)
+    config.model_spec.converters.append(lora_converter)
```
Comment (tianyu-l) [inline torchtitan/models/llama3/config_registry.py]: we should extend the `quantization` field in `model_registry` from quantization converters to general converters, and apply converters in sequential order
https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/llama3/config_registry.py#L100-L102

A safety check is that LoRAConverters have to be applied after all quantization converters.

File: torchtitan/components/lora.py (hunk 6 above)
Comment (tianyu-l) [reply torchtitan/components/lora.py]: also, with that I don't think you still need this `LoRAConfig` class.

**Review by tianyu-l — COMMENTED** [review]
Only reviewed the LoRA logic. Will review state dict transformation later.

File: torchtitan/components/lora.py (hunk 7 above)
Comment (mori360) [reply torchtitan/components/lora.py:129]: Have the MXFP8GroupedExperts pr here: https://github.com/pytorch/torchtitan/pull/3251

File: torchtitan/models/qwen3_vl/__init__.py [hunk 9]
```diff
@@ -591,4 +595,5 @@ def model_registry(
         pipelining_fn=None,
         post_optimizer_build_fn=None,
         state_dict_adapter=Qwen3VLStateDictAdapter,
+        converters=built_converters,
```
Comment (tianyu-l) [inline torchtitan/models/qwen3_vl/__init__.py:598]: don't need this in `ModelSpec`?

File: torchtitan/models/qwen3/__init__.py [hunk 10]
```diff
@@ -619,15 +619,19 @@ def model_registry(
...
         kwargs["moe_comm_backend"] = moe_comm_backend
     config = qwen3_configs[flavor](**kwargs)
-    if quantization is not None:
-        for q in quantization:
-            q.build().convert(config)
+    built_converters: list = []
+    if converters is not None:
+        for c in converters:
+            built_converters.append(c.build())
+        validate_converter_order(built_converters)
```
Comment (tianyu-l) [inline torchtitan/models/qwen3/__init__.py:632]: I think can validate before `c.build()` -- just need to validate on config, e.g. `if isinstance(converter, LoRAConverter.Config):`

File: torchtitan/protocols/model_spec.py [hunk 11]
```diff
@@ -23,6 +23,28 @@
 PostOptimizerBuildFn: TypeAlias = Callable[..., None]
 
 
+def validate_converter_order(converters: list) -> None:
```
Comment (tianyu-l) [inline torchtitan/protocols/model_spec.py:26]: let's put in torchtitan/models/utils.py for now

**Review by tianyu-l — COMMENTED** [review]
Right now the lora PR is after checkpointing PR. I think it's easier to land 1. LoRA first, then 2. checkpointing, then 3. lora + checkpointing (can be combined with 2).

I.e. it should be fine to have a version of LoRA which doesn't work with checkpointing compatible.
